### PR TITLE
Update `failure_type_classifier`

### PR DESCRIPTION
**Pull request by isabelizimm**
Co-authored-by: Shrey Anand <shanand@redhat.com>

## Related Issues and Dependencies
#274 

## This introduces a breaking change

- [ ] Yes
- [X] No

## This Pull Request implements
code for running the failure type classification in a pipeline.

## Description

- Changed detect-install-flake to account for the edge case where there aren't any.
- Changed detect-new-test-failure to account for reference before assignment error.
- Tested the code with data from these dates: 810, 185, 36, 165, 144, 76, 86
- Link to running pipeline

Things that can be improved here:
- Define an output schema based on how it'll be consumed (currently it's a json file)
- Fix errors for some dashboards and grids (currently it's been skipped)
- Change output paths based on where we want this stored
- Dependencies have to be in the same directory which is a limitation of Elyra and so we have two copies of metric_template.ipynb for now. Upstream issue.

### CODE REVIEW

File: notebooks/failure-type-classification/stage/fail_type_functions.ipynb [hunk 1]
```diff
@@ -0,0 +1,343 @@
+{
```
Comment (MichaelClifford) [inline notebooks/failure-type-classification/stage/fail_type_functions.ipynb:1]: <p>Can we move all of these out of <code>stage</code>?  and update all the paths accordingly? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: notebooks/failure-type-classification/stage/fail_type_functions.ipynb [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    "    infra_flake_filter = np.array([[-1, 1], [1, -1]])\n",
+    "\n",
+    "    # Find the spots on the map where the convolution had its maximum value.\n",
+    "    spots = convolve2d(infra_flake_filter, grid, mode=\"valid\")\n",
+    "    infra_flakes = np.where(spots == 4)\n",
+    "\n",
+    "    dates = data[tab_name][grid_name][\"timestamps\"]\n",
+    "    infra_flake_dates = np.array(dates)[list([infra_flakes][0][1])]\n",
+    "    infra_flake_dates = [\n",
+    "        datetime.date.fromtimestamp(x // 1000) for x in infra_flake_dates #issue with x here...what is it referring to?\n",
```
Comment (MichaelClifford) [inline notebooks/failure-type-classification/stage/fail_type_functions.ipynb:90]: @isabelizimm 
I did some testing with pre-commit and I think the issue is with the comment in this line. Removing it seems to fix the issue. 

Also, to answer the comment so that it can be removed, this is a list comprehension; works similar to a for loop in the form [ do_something_with_(x) for x in iterator] and returns a list. In this case we are converting items in infra_flake_dates into datetime format from timestamps.

Comment (isabelizimm) [reply notebooks/failure-type-classification/stage/fail_type_functions.ipynb:90]: Ahhh yes. When I rearranged the functions I had to update a few things, but forgot I had put that comment in there. When I removed this comment, my pre-commit worked as well. Thank you for doing some detective work!

File: notebooks/failure-type-classification/failure.pipeline [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+              "S3_ACCESS_KEY=v3FnruQ78kfeULDjejUB",
+              "S3_SECRET_KEY=kJiDiHXncLJOXbaL7Zeb5Ok+gkLt9sWIa1rWAJa0",
+              "S3_BUCKET=opf-datacatalog",
+              "IN_AUTOMATION=True"
+            ],
+            "include_subdirectories": false,
+            "invalidNodeError": null,
+            "outputs": [],
+            "dependencies": [
+              "metric_template.ipynb",
```
Comment (MichaelClifford) [inline notebooks/failure-type-classification/failure.pipeline:29]: Can you update this line to point to the correct location of 'notebooks/data-sources/TestGrid/metrics/metric_template.ipynb' and see if this works instead of duplicating the notebook in this dir?

File: notebooks/failure-type-classification/failure_type_classifier.ipynb [hunk 4]
```diff
@@ -27,7 +27,7 @@
   },
```
Comment (MichaelClifford) [inline notebooks/failure-type-classification/failure_type_classifier.ipynb:27]: <pre class="ql-syntax" spellcheck="false">Line #7.        filename = f"report_local.json"
</pre><p>Should the locally saved file also have the same day.month.dashboard.job convention?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

File: notebooks/failure-type-classification/failure_type_functions.ipynb [hunk 5]
```diff
@@ -0,0 +1,328 @@
+{
```
Comment (MichaelClifford) [inline notebooks/failure-type-classification/failure_type_functions.ipynb:1]: <p>Can you update this markdown to reflect this notebook is just the helper functions? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: notebooks/failure-type-classification/metric_template.ipynb [hunk 6]
```diff
@@ -0,0 +1,973 @@
+{
```
Comment (MichaelClifford) [inline notebooks/failure-type-classification/metric_template.ipynb:1]: <p>If we there is no option to call this notebook from its original location, can we customize it to be "helper for failure type" so it is its own notebook? And remove anything that's not needed for the workflow here? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: notebooks/failure-type-classification/testgrid_flakiness_detection.ipynb [hunk 7]
```diff
@@ -18,7 +18,7 @@
   },
```
Comment (MichaelClifford) [inline notebooks/failure-type-classification/testgrid_flakiness_detection.ipynb:18]: <p>This can be included in the top cell. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

File: notebooks/failure-type-classification/failure.pipeline (hunk 3 above)
Comment (isabelizimm) [reply notebooks/failure-type-classification/failure.pipeline:29]: This was duplicated for the pipeline dependencies. For the pipeline to run without error, it needs the notebooks that are installing the dependencies, rather than just references to other notebooks that install the dependencies. This is the current workaround, lmk if there is a better way to make this work :)

Comment (MichaelClifford) [reply notebooks/failure-type-classification/failure.pipeline:29]: :+1: Yes, I realized its not that simple after I made the comment. That said, I am pretty opposed to duplicating notebooks across the repo as it is an unsustainable pattern. We are looking into some possible alternatives now, but if none of them pan out we can apply this temporarily until https://github.com/elyra-ai/elyra/pull/1689 is included into the latest release.
